src/jdk.internal.le/share/classes/jdk/internal/org/jline/JdkConsoleProviderImpl.java

```diff
@@ -0,0 +1,120 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.org.jline;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.Reader;
+import java.io.UncheckedIOException;
+import java.nio.charset.Charset;
+
+import jdk.internal.io.JdkConsole;
+import jdk.internal.io.JdkConsoleProvider;
+import jdk.internal.org.jline.reader.EndOfFileException;
+import jdk.internal.org.jline.reader.LineReader;
+import jdk.internal.org.jline.reader.LineReaderBuilder;
+import jdk.internal.org.jline.terminal.Terminal;
+import jdk.internal.org.jline.terminal.TerminalBuilder;
+
+/**
+ * JdkConsole/Provider implementations for jline
+ */
+public class JdkConsoleProviderImpl implements JdkConsoleProvider {
+
+    /**
+     * {@inheritDoc}
+     */
+    public JdkConsole console(boolean isTTY) {
+        return new JdkConsoleImpl();
+    }
+
+    /**
+     * An implementation of JdkConsole, which act as a delegate for the
+     * public Console class.
+     */
+    public static class JdkConsoleImpl implements JdkConsole {
+        public PrintWriter writer() {
+            return terminal.writer();
+        }
+
+        public Reader reader() {
+            return terminal.reader();
+        }
+
+        public JdkConsole format(String fmt, Object ... args) {
+            writer().format(fmt, args).flush();
+            return this;
+        }
+
+        public JdkConsole printf(String format, Object ... args) {
+            return format(format, args);
+        }
+
+        public String readLine(String fmt, Object ... args) {
+            try {
+                return jline.readLine(fmt.formatted(args));
+            } catch (EndOfFileException eofe) {
+                return null;
+            }
+        }
+
+        public String readLine() {
+            return readLine("");
+        }
+
+        public char[] readPassword(String fmt, Object ... args) {
+            try {
+                return jline.readLine(fmt.formatted(args), '\0').toCharArray();
+            } catch (EndOfFileException eofe) {
+                return null;
+            }
+        }
+
+        public char[] readPassword() {
+            return readPassword("");
+        }
+
+        public void flush() {
+            terminal.flush();
+        }
+
+        public Charset charset() {
+            return terminal.encoding();
+        }
+
+        private final LineReader jline;
+        private final Terminal terminal;
+
+        public JdkConsoleImpl() {
+            try {
+                terminal = TerminalBuilder.builder().build();
```

Comment:
The `java.io.Console` in its static initialization code has some logic to determine the `Charset` to use. Should that same `Charset` (or logic) be used here to build the terminal? Something like `TerminalBuilder.builder().encoding(fooBarCharset).build();`.